feat: implemented required by Task scope-methods and validations

SergeSadowsky opened the pull request:
1. Implemented scope-methods:

- Answer.correct
- Test.simple
- Test.average
- Test.difficult
- default Category ordered ascending

2. Refactored existing methods of User and Test using scope-methods 

3. Created validations for:
title, body, email attributes of corresponding classes
Test.level must be integer, greater than 0
Uniqueness title, level attributes of Test class
Question can't have more than 4 answers

uncleserge committed:
feat: implemented required by Task scope-methods and validations

BubuntuClu commented on app/models/test.rb:
```diff
@@ -4,9 +4,22 @@ class Test < ApplicationRecord
+  scope :average, -> { level(2..4) }
+  scope :difficult, -> { level(5..Float::INFINITY) }
+
+  scope :by_category, -> (category) { joins(:category).where(categories: {title: category}).pluck(:title) }
```
сортировку и выборку не надо помещать в скоуп.
скоуп должен возвращать объект ActiveRecord::Relation, pluck возвращает массив.
так же скоуп это универсальный метод, который может использован в разных частях приложения, там может потребоваться другая сортировка и выборка.
метод надо оставить, заменить часть выражения на скоуп и уже к скоупу применить сортировку и выборку

SergeSadowsky replied:
Fixed

BubuntuClu commented on app/models/user.rb:
```diff
@@ -4,7 +4,13 @@ class User < ApplicationRecord
+  validates :first_name, presence: true
+  validates :last_name, presence: true
+
+  scope :tests_by_level, -> (level) { tests.where(level: level) }
```
проверял этот код? он работает так же как и метод, который заменил?

SergeSadowsky replied:
Fixed

BubuntuClu commented on app/models/test.rb:
```diff
@@ -4,9 +4,22 @@ class Test < ApplicationRecord
-    Test.joins(:category).where(categories: {title: category}).pluck(:title)
-  end
+
+  validates :title, presence: true,
```
ограничения надо дублировать на всех уровнях приложения. можно через СУБД насоздавать данных, которые для приложения не будут валидны. в БД добавить уникальный составной индекс надо

SergeSadowsky replied:
Fixed

uncleserge committed:
fix: implemented corrections for task:
Test.by_category
User.tests_by_level
add migration with compound index for Title, level attributes of Test class

BubuntuClu commented on app/models/user.rb:
```diff
@@ -4,7 +4,16 @@ class User < ApplicationRecord
+  validates :first_name, presence: true
+  validates :last_name, presence: true
+
+  # def tests_by_level(level)
```
ненужный код удаляй, не надо его тащить комментами, он будет мешать и генерировать вопросы "а зачем он тут?". вырабатывай привычку подчищать за собой после рефакторинга. различные изменения в файле можно в истории глянуть на гитхабе

BubuntuClu commented:
всю папку `idea` надо добавить в гитигнор. другим разработчикам твои настройки окружения не нужны. в проекте должен быть код только по проекту

uncleserge committed:
fix: deleted comments and .idea went to gitignore
